src/gt4py/analysis/passes.py

```diff
@@ -1210,6 +1210,95 @@ def apply(self, transform_data: TransformData) -> TransformData:
         return transform_data
 
 
+class ReduceTemporaryStoragesPass(TransformPass):
+    class StorageReducer(gt_ir.IRNodeMapper):
+        def __init__(self):
+            self.iir = None
+            self.interval = None
+            self.iteration_order = None
+            self.full_fields = set()
+            self.reduced_fields = dict()
+            self.fields_extents = dict()
+
+        def __call__(self, node: gt_ir.StencilImplementation) -> gt_ir.StencilImplementation:
+            assert isinstance(node, gt_ir.StencilImplementation)
+            result = self.visit(node)
+            self._reduce_fields()
+            return result
+
+        def _reduce_fields(self):
+            for field_name in self.reduced_fields:
+                field_decl = self.iir.fields[field_name]
+                field_decl.axes.pop()
+                for node in self.reduced_fields[field_name]["nodes"]:
+                    new_offset = {axis: node.offset[axis] for axis in field_decl.axes}
+                    node.offset = new_offset
+
+        def visit_StencilImplementation(
+            self, path: tuple, node_name: str, node: gt_ir.StencilImplementation
+        ) -> gt_ir.StencilImplementation:
+            self.iir = node
+            self.fields_extents = node.fields_extents.copy()
+            return self.generic_visit(path, node_name, node)
+
+        def visit_MultiStage(
+            self, path: tuple, node_name: str, node: gt_ir.MultiStage
+        ) -> Tuple[bool, Optional[gt_ir.MultiStage]]:
+            self.iteration_order = node.iteration_order
+            self.generic_visit(path, node_name, node)
+            return True, node
+
+        def visit_ApplyBlock(
+            self, path: tuple, node_name: str, node: gt_ir.ApplyBlock
+        ) -> Tuple[bool, gt_ir.ApplyBlock]:
+            self.interval = node.interval
+            self.generic_visit(path, node_name, node)
+            return True, node
+
+        def visit_FieldAccessor(
+            self, path: tuple, node_name: str, node: gt_ir.FieldAccessor
+        ) -> Tuple[bool, Optional[gt_ir.FieldAccessor]]:
+            self.fields_extents[node.symbol] |= node.extent
+            return True, node
+
+        def visit_FieldRef(
+            self, path: tuple, node_name: str, node: gt_ir.FieldRef
+        ) -> Tuple[bool, gt_ir.FieldRef]:
+            field_name = node.name
+            is_full_field = False
+            if field_name in self.iir.temporary_fields:
+                if self.iteration_order == gt_ir.IterationOrder.PARALLEL:
+                    is_full_field = True
+
+                elif field_name not in self.full_fields:
+                    extent = self.fields_extents[field_name]
+                    ndims = extent.ndims - 1
+                    if extent.lower_indices[ndims] == 0 and extent.upper_indices[ndims] == 0:
```

Comment:
I don't think this is a good condition. the read/write pattern should be such that the value of the previous index (i.e. -1 if FORWARD else 1) is read, while the current is written (and possibly read after). Further, the field must never be read outside of the same stage (with some tricky corner cases allowed, such as e.g. outside of that stage only reading the last written layer)